src/AutomaticScheduling.cpp

```diff
@@ -0,0 +1,268 @@
+    vector<Function> callers = cg.callers(f);
+    if (callers.empty()) {
+        // Return undefined bounds for uncalled functions.
+        result.clear();
+        return result;
+    }
+    for (vector<Function>::const_iterator I = callers.begin(), E = callers.end(); I != E; ++I) {
+        const vector<Expr> outputs = (*I).values();
+        internal_assert(outputs.size() == 1) << "Unhandled number of outputs";
+        Box b = boxes_required(outputs[0])[f.name()];
+        const unsigned dim = b.bounds.size();
+        if (dim == 0) {
+            // Unable to compute the bounds for the function. Return undefined bounds.
+            result.clear();
+            return result;
+        }
+        internal_assert((int)dim == f.dimensions());
+        for (unsigned i = 0; i < dim; ++i) {
+            Interval interval = b.bounds[i];
+            if (!initialized[i]) {
+                initialized[i] = true;
+                result[i] = interval;
+            } else {
+                result[i].min = min(result[i].min, interval.min);
+                result[i].max = max(result[i].max, interval.max);
+            }
+        }
+    }
+    return result;
+}
+
+
+/** Return the footprint (required region) of the given function over
+ * all callsites. This is a minimum of 1 when a function is
+ * pointwise. If the footprint cannot be calculated, returns
+ * UNDEFINED_FOOTPRINT_SIZE. */
+unsigned calculate_footprint_size(Function f, CallGraph &cg) {
+    vector<Interval> bounds = get_function_bounds(f, cg);
+    if (bounds.size() == 0) {
+        // Unable to calculate function bounds.
+        return UNDEFINED_FOOTPRINT_SIZE;
+    }
+    Expr footprint = 1;
+    for (vector<Interval>::iterator I = bounds.begin(), E = bounds.end(); I != E; ++I) {
+        const Interval &i = *I;
+        internal_assert(i.min.defined() && i.max.defined());
+        Expr diff = i.max - i.min + 1;
+        footprint *= diff;
+    }
+    const int *result = as_const_int(simplify(footprint));
+    if (result) {
+        return (unsigned)*result;
+    } else {
+        return UNDEFINED_FOOTPRINT_SIZE;
+    }
+}
+
+} // end anonymous namespace
+
+void ComputeRootAllStencils::apply(Func root) {
+    // Construct a callgraph for the pipeline.
+    CallGraph cg(root.function());
+    vector<Function> all_functions = cg.transitive_calls(root.function());
+    for (vector<Function>::iterator I = all_functions.begin(), E = all_functions.end(); I != E; ++I) {
+        Function f = *I;
+        unsigned footprint = calculate_footprint_size(f, cg);
+        if (footprint > 1) {
+            Func wrapper(f);
+            wrapper.store_root().compute_root();
+        }
+    }
+}
+
+void ParallelizeOuter::apply(Func root) {
+    CallGraph cg(root.function());
+    vector<Function> all_functions = cg.transitive_calls(root.function());
+    all_functions.push_back(root.function());
+    for (vector<Function>::iterator I = all_functions.begin(), E = all_functions.end(); I != E; ++I) {
```

Comment:
Either use a range-based for loop for these loops over vectors, or just use:

for (size_t i = 0; i < all_functions.size(); i++) {
  Function f = all_functions[i];
  ...
}

which is the idiom used elsewhere in the Halide source.